### PR TITLE
Fix Store API response for very large price amounts

### PR DESCRIPTION
**Pull request by wavvves**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR addresses improper formatting for very high price amounts on Store API responses

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Set up a store using Euro currency and 2 decimal points (options on `/wp-admin/admin.php?page=wc-settings`)
2. Disable taxes.
3. Add a new product with the price `9223372036854775808` and take note of the product id
4. Add that product to cart
5. Verify that on `/wp-json/wc/store/v1/cart` response JSON the `items[0].prices.price` value is correctly displayed as  `922337203685477580800`
6. Verify that on `/wp-json/wc/store/v1/products/<product id>` response JSON `prices.price` value is correctly displayed as  `922337203685477580800`

Please note that totals calculations on Cart and Checkout is not properly done. This is not a consequence of this PR and will be addressed on a different issue. Same with Cart/Checkout blocks wrong rounding for values, which is a limitation from the Dinerojs npm package.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php [hunk 1]
```diff
@@ -23,12 +24,24 @@ public function format( $value, array $options = [] ) {
...
-			)
-		);
+		// Ensure rounding mode is valid.
+		$rounding_modes           = [ PHP_ROUND_HALF_UP, PHP_ROUND_HALF_DOWN ];
+		$options['rounding_mode'] = absint( $options['rounding_mode'] );
+		if ( ! in_array( $options['rounding_mode'], $rounding_modes, true ) ) {
+			$options['rounding_mode'] = PHP_ROUND_HALF_UP;
+		}
+
+		// This ensures we work with floats from this point on. Also, if $value skips this as an array/object,
```
Comment (opr) [inline plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: Do you think explaining _why_ we need to work with floats would be beneficial?

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php [hunk 2]
```diff
@@ -23,12 +24,24 @@ public function format( $value, array $options = [] ) {
...
+		// it will throw a fatal error or produce an E_NOTICE (if Object on PHP <8.0).
```
Comment (opr) [inline plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: Curious about the fatal, is it OK fatal here or should we try to mitigate somehow?

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php [hunk 3]
```diff
@@ -23,12 +24,24 @@ public function format( $value, array $options = [] ) {
...
 
-		return (string) intval(
-			round(
-				( (float) wc_format_decimal( $value ) ) * ( 10 ** absint( $options['decimals'] ) ),
-				0,
-				absint( $options['rounding_mode'] )
-			)
-		);
+		// Ensure rounding mode is valid.
+		$rounding_modes           = [ PHP_ROUND_HALF_UP, PHP_ROUND_HALF_DOWN ];
```
Comment (barryhughes) [inline plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: I think this is probably fine, but since it's effectively part of the public API surface, I'll ask anyway: is there any risk to dropping support for `PHP_ROUND_HALF_EVEN` and `PHP_ROUND_HALF_ODD`?

These were previously allowed and supported in the earlier version of the method.

**Review by barryhughes — APPROVED** [review]
Looks good to me; left a question, but we may not need to change anything (I was partly just curious).

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php (hunk 3 above)
Comment (wavvves) [reply plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: Makes total sense! I started by having those too and removed them without thinking it was previously supported after seeing only these scattered around 🫠
I will re-add them 👍

**Review by opr — APPROVED** [review]
Working for me, thanks, just two minor comments for you to address before merging.

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php (hunk 2 above)
Comment (wavvves) [reply plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: My rationale for this was avoiding converting arrays/objects to floats since that can result in the 1/0 value prices. But since the original param type was soft declared as mixed, and this is a public API you are right about this.

Inspecting the original code:

`( (float) wc_format_decimal( $value ) ) * ( 10 ** absint( $options['decimals'] ) ),`

This is where $value is used, consumed by `wc_format_decimal()`. First use is with:

```php
if ( ! is_float( $number ) ) {
    $number = str_replace( $decimals, '.', $number );

    // Convert multiple dots to just one.
    $number = preg_replace( '/\.(?![^.]+$)|[^0-9.-]/', '', wc_clean( $number ) );
}
```

And `str_replace`, `preg_replace` and `wc_clean` accept arrays. No fatal so far.

All the next if clauses would be avoided and it would return an array, which would be converted to `1` or '0' after the (float) typecast.

It would fatal with another param type passed in. 

So to sum up I'm inclined to add `is_array()` to the if and recheck params for `wc_format_decimal`, to keep the same behaviour regarding fatals/wrong value prices due to array > float conversions, and add a doing_it_wrong if the param is not the one newly listed. And remove mixed from the method sig.

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php [hunk 4]
```diff
@@ -23,12 +24,30 @@ public function format( $value, array $options = [] ) {
...
+
+		if ( ! is_int( $value ) && ! is_string( $value ) && ! is_float( $value ) ) {
+			wc_doing_it_wrong(
+				__FUNCTION__,
+				__( 'This function expects a $value arg of type INT, STRING or FLOAT.', 'woocommerce' ),
+				'9.2'
+			);
+
+			return '';
+		}
```
Comment (barryhughes) [inline plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: Definitely not essential we change this, but might it make sense to move this higher (that is, perform this as our first check before doing any other work)?

I'm not suggesting this as a micro-optimization, because we're not really doing anything super expensive above this point, but it feels (to me) like a slightly more logical (and readable) way to order things.

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php [hunk 5]
```diff
@@ -23,12 +24,30 @@ public function format( $value, array $options = [] ) {
...
+		$rounding_modes           = [ PHP_ROUND_HALF_UP, PHP_ROUND_HALF_DOWN, PHP_ROUND_HALF_EVEN, PHP_ROUND_HALF_ODD ];
+		$options['rounding_mode'] = absint( $options['rounding_mode'] );
+		if ( ! in_array( $options['rounding_mode'], $rounding_modes, true ) ) {
+			$options['rounding_mode'] = PHP_ROUND_HALF_UP;
+		}
+
+		if ( ! is_int( $value ) && ! is_string( $value ) && ! is_float( $value ) ) {
+			wc_doing_it_wrong(
+				__FUNCTION__,
+				__( 'This function expects a $value arg of type INT, STRING or FLOAT.', 'woocommerce' ),
```
Comment (barryhughes) [inline plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: I'm not sure we should be translating this (though, yes, I know there are many existing places in the code base where we *do* translate logging messages).

I'm thinking here of our [logging best practices](https://github.com/woocommerce/woocommerce/blob/9.1.1/docs/extension-development/logging.md#best-practices) (though mostly written in relation to `wc_get_logger()`, I think many of the principles carry over to logging in general).

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php [hunk 6]
```diff
@@ -23,12 +24,30 @@ public function format( $value, array $options = [] ) {
...
-		);
+		// Ensure rounding mode is valid.
+		$rounding_modes           = [ PHP_ROUND_HALF_UP, PHP_ROUND_HALF_DOWN, PHP_ROUND_HALF_EVEN, PHP_ROUND_HALF_ODD ];
+		$options['rounding_mode'] = absint( $options['rounding_mode'] );
+		if ( ! in_array( $options['rounding_mode'], $rounding_modes, true ) ) {
+			$options['rounding_mode'] = PHP_ROUND_HALF_UP;
+		}
+
+		if ( ! is_int( $value ) && ! is_string( $value ) && ! is_float( $value ) ) {
+			wc_doing_it_wrong(
```
Comment (barryhughes) [inline plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: Another non-essential suggestion (feel free to discard), we could describe this behaviour with an addition test or assertion.

```php
$this->assertEquals( 
	'', 
	$this->mock_formatter->format( array( 9, 99 ) ),
	'When an unexpected type such as an array is passed, the result will be an empty string.'
);
```

(☝🏼 Not tested, just illustrative.)

**Review by barryhughes — APPROVED** [review]
Nice! Left some suggestions, but they are non-essential and we can proceed without them.

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php (hunk 6 above)
Comment (barryhughes) [reply plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: Or perhaps `$this->expectDeprecated()` ...

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php (hunk 5 above)
Comment (opr) [reply plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: Agree with this, especially since developers would likely search the error online, having different errors per language reduce the chances of them finding a relevant discussion/fix.

**Review by opr — APPROVED** [review]
I'd suggest modifying the test instructions to mention:

- Disable tax (this makes it easier to verify the expected value in the next steps)
- Be more explicit about which key we should check in the JSON response
- Add some non-technical steps, e.g. how can this be tested on the front-end only without looking at JSON.

Besides these suggestions, and the ones Barry mentioned, this is working well, thanks Paulo.

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php (hunk 4 above)
Comment (wavvves) [reply plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: 😅 indeed

done in https://github.com/woocommerce/woocommerce/pull/49361/commits/3767dedfcb114689066e4ac5107ddb806efa11e0

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php (hunk 5 above)
Comment (wavvves) [reply plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: I added it because, indeed, it appears often throughout our codebase.

removed in https://github.com/woocommerce/woocommerce/pull/49361/commits/3767dedfcb114689066e4ac5107ddb806efa11e0

File: plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php (hunk 6 above)
Comment (wavvves) [reply plugins/woocommerce/src/StoreApi/Formatters/MoneyFormatter.php]: Actually, I did add tests for this, but I'm glad you brought this up because I forgot to rename the test function.

**Comment by wavvves** [comment]
> I'd suggest modifying the test instructions to mention:
> 
> * Disable tax (this makes it easier to verify the expected value in the next steps)
> * Be more explicit about which key we should check in the JSON response
> * Add some non-technical steps, e.g. how can this be tested on the front-end only without looking at JSON.
> 
> Besides these suggestions, and the ones Barry mentioned, this is working well, thanks Paulo.

@opr I added your suggestions to the testing steps, but not sure how to check this without looking at API responses. We currently have a Dinerojs npm package limitation on blocks that will round that value by 4 digits, and the value presented on shortcode doesn't make use of this method 🤔
